crypto/x509/x509_vfy.c

```diff
@@ -973,10 +992,16 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
 static int check_revocation(X509_STORE_CTX *ctx)
 {
     int i = 0, last = 0, ok = 0;
+    int crl_ocsp_check =
+        ctx->param->flags & (X509_V_FLAG_CRL_CHECK | X509_V_FLAG_OCSP_RESP_CHECK);
+    int crl_ocsp_check_all =
+        ctx->param->flags
+        & (X509_V_FLAG_CRL_CHECK_ALL | X509_V_FLAG_OCSP_RESP_CHECK_ALL);
```

Comment:
One can not combine checking the flags `X509_V_FLAG_CRL_CHECK_ALL` and `X509_V_FLAG_OCSP_RESP_CHECK_ALL` this way,
as it may be that just one of them is set while both 
`X509_V_FLAG_CRL_CHECK` and `X509_V_FLAG_OCSP_RESP_CHECK` are set.
In this case, your below code checks cert status for the full chain with both OCSP responses and CRLs, which is not what was requested.

(Alternatively, one might introduce `X509_V_FLAG_STATUS_CHECK{,_ALL}` in a way that they mean 
not only `X509_V_FLAG_CRL_CHECK{,_ALL}`, but also any further types of status checks, including the OCSP-based ones implemented here.)

Reply:
Currently the OCSP status checking is only implemented for the app `s_client`. I could extend it also to the general `verify`paramters and add a parameter `-ocsp_check`. 
In that case I can change the behaviour of the applications that the parameter `-crl_check_all` is deprecated and there is a new parameter `-check_full_chain` instead. 
Or I leave the parameters `-crl_check` and `-crl_check_all` untouched and add the parameter `-check_full_chain`, in this case at least the combination of `-crl_check_all` and `-ocsp_check` will still result in the checking of the full chain.

What does you mean?

Reply:
Yeah, unfortunately the need for semantic backward compatibility and the fact that the new implementation only covers part of all possible OCSP-based checks complicates matters. 
This is why I put the alternative that I mentioned in parentheses.

Something like `-check_full_chain` or more precisely `-check_status_all` would be great -
yet would ask for a full implementation, 
including client-side stapling and general OCSP response fetching on either side.

BTW, good that you reminded me that this PR just covers the client-side checks of multi-stapling.
So, unless the OCSP support get strongly generalized, I fear you have to rename your status flags to, e.g., 
`X509_V_FLAG_CHECK_SERVER_OCSP_RESP` and 
`X509_V_FLAG_CHECK_SERVER_OCSP_RESP_ALL`
and use in the implementation two different `last` variables/values, one for OCSP stapling and one for CRLs.

Reply:
Let me check what I can do and implement the OCSP checking in general.
The point with the `last` variable would not work, because if both CRL and OCSP is enabled, CRLs are only checked if OCSP was not working before.

Reply:
Hello @DDvO, hello @t8m, hello @FdaSilvaYY,

please ignore my previous post. I read further the RFC 8446 for TLS 1.3 and found that there is the option that the server requests OCSP stapling from the client.

The current implementation is not using the CertificateStatusType  `ocsp_multi(2)` as defined in RFC 6961. My suggestion would be that for the app s_client the `status` command line option will remain for single staplign and I would add an option `status_multi` (instead of `ocsp_check_leaf` and `ocsp_check_all`). 
In case of `status` only the leaf certificate should be checked, as only for it the OCSP response is provided by the server. In the case of `status_multi` the complete chain is checked.

For the server side I would suggest, first to finish this PR for the client. Then I or somebody else could have a look at the server side. Would that be ok?

Best regard,
Martin

Reply:
> Give me some time. I have to think about, what would be the best solution.

Fine!

Reply:
Hello @DDvO,

here are the result of my thoughts:

**No OCSP Flag**
- No CRL
  - No checks

- X509_V_FLAG_CRL_CHECK
  - CRL check of leaf certificate   

- X509_V_FLAG_CRL_CHECK_ALL
  - CRL check of chain certificates

**X509_V_FLAG_OCSP_RESP_CHECK**
- No CRL
  - OCSP check of leaf certificate
		
- X509_V_FLAG_CRL_CHECK
  - OCSP check of leaf certificate
  - CRL check of leaf cerrtificate will only be checked if OCSP Response is not provided (X509_V_ERR_OCSP_NO_RESPONSE)
		
- X509_V_FLAG_CRL_CHECK_ALL		
  - OCSP check of leaf certificate
  - CRL check of leaf cerrtificate will only be checked if OCSP Response is not provided (X509_V_ERR_OCSP_NO_RESPONSE)
  - CRL check for the rest of the chain certificates

**X509_V_FLAG_OCSP_RESP_CHECK_ALL**
- No CRL: 
  - OCSP check of chain certificates
- X509_V_FLAG_CRL_CHECK: 
  - OCSP check of chain certificates
  - CRL check of leaf cerrtificate will only be checked if OCSP Response is not provided (X509_V_ERR_OCSP_NO_RESPONSE)

- X509_V_FLAG_CRL_CHECK_ALL
  - OCSP check of chain certificates
  - CRL check of chain certificates will only be checked if OCSP Response is not provided (X509_V_ERR_OCSP_NO_RESPONSE)

Would that be ok to implement the checks in that way?

As I wrote I would postpone the server side checks.

Reply:
Hello @martinRa2, 
this sounds very good, provided that with 
"if OCSP Response is not provided  (X509_V_ERR_OCSP_NO_RESPONSE)"
you actually mean "OCSP check was inconclusive", which boils down to any other result than
V_OCSP_CERTSTATUS_REVOKED or V_OCSP_CERTSTATUS_GOOD.

Since you are not going to implement the client-side (multi-)stapling,
return an error if X509_V_FLAG_OCSP_RESP_CHECK or X509_V_FLAG_OCSP_RESP_CHECK_ALL are set on server side
and point out clearly in the API doc that so far only server-side (multi-)stapling is supported.